[GTK][WPE] Add support for EGL fences

carlosgcampos opened the pull request:
#### a258d90c0cadfb539a4472069a4874ecd646ac64
<pre>
[GTK][WPE] Add support for EGL fences
<a href="https://bugs.webkit.org/show_bug.cgi?id=277141">https://bugs.webkit.org/show_bug.cgi?id=277141</a>

Reviewed by Miguel Gomez.

For explicit sync support we need to use the EGL API to create the
fences to be able to use the EGL_ANDROID_native_fence_sync extension.
So, we can make it possible to create a GLFence with either gl or egl
APIs depending on what we need and what&apos;s available.
There&apos;s no change in behavior in this patch, but now that we have EGL
fences we prefer to use them when available if server wait is supported.

* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::initializeEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/egl/GLFence.cpp:
(WebCore::GLFence::capabilities):
(WebCore::GLFence::isSupported):
(WebCore::GLFence::create):
(WebCore::GLFence::createExportable):
(WebCore::GLFence::importFD):
(WebCore::GLFence::GLFence): Deleted.
(WebCore::GLFence::~GLFence): Deleted.
(WebCore::GLFence::wait): Deleted.
* Source/WebCore/platform/graphics/egl/GLFence.h:
(WebCore::GLFence::exportFD):
* Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp: Added.
(WebCore::createEGLFence):
(WebCore::GLFenceEGL::create):
(WebCore::GLFenceEGL::createExportable):
(WebCore::GLFenceEGL::importFD):
(WebCore::GLFenceEGL::GLFenceEGL):
(WebCore::GLFenceEGL::~GLFenceEGL):
(WebCore::GLFenceEGL::clientWait):
(WebCore::GLFenceEGL::serverWait):
(WebCore::GLFenceEGL::exportFD):
* Source/WebCore/platform/graphics/egl/GLFenceEGL.h: Added.
* Source/WebCore/platform/graphics/egl/GLFenceGL.cpp: Added.
(WebCore::GLFenceGL::create):
(WebCore::GLFenceGL::GLFenceGL):
(WebCore::GLFenceGL::~GLFenceGL):
(WebCore::GLFenceGL::clientWait):
(WebCore::GLFenceGL::serverWait):
* Source/WebCore/platform/graphics/egl/GLFenceGL.h: Added.

Canonical link: <a href="https://commits.webkit.org/281558@main">https://commits.webkit.org/281558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc735cbaee5a4ace18a1107bcdd5b3942db27cf5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64255 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10867 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62465 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11100 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7568 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62366 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/36992 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/52274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29697 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33690 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/9500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9784 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/9787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65987 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4269 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56211 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4287 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/52247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56379 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/3559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9045 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37668 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

magomez commented on Source/WebCore/PlatformPlayStation.cmake:
```diff
@@ -99,7 +99,11 @@ if (ENABLE_WEBGL)
-    list(APPEND WebCore_SOURCES platform/graphics/egl/GLFence.cpp)
+    list(APPEND WebCore_SOURCES
+        platform/graphics/egl/GLFence.cpp
+        platform/graphics/egl/GLFenceARB.cpp
```
Why GLFenceARB? We're not using any ARB extension here. That fence will be using the GLES3 API, so it would make more sense as GLFenceGL, following the pattern of GLFenceEGL.

carlosgcampos replied:
I agree, I've renamed it.

magomez commented on Source/WebCore/platform/graphics/egl/GLFence.cpp:
```diff
@@ -21,70 +21,87 @@
+#if OS(LINUX)
+        capabilities.eglExportableSupported = extensions.ANDROID_native_fence_sync;
+#endif
+        capabilities.glSupported = GLContext::versionFromString(reinterpret_cast<const char*>(glGetString(GL_VERSION))) >= 300;
```
We need gl fence support in order to have egl fence support, so the values of eglSupported and eglServerWaitSupported should depend on the value of glSupported.

carlosgcampos replied:
I assume this is implicit, if driver exposes egl fence extensions is because we can use the api, no? If it fails later at runtime to create the fence it will have the same effect.

magomez commented on Source/WebCore/platform/graphics/egl/GLFence.cpp:
```diff
@@ -21,70 +21,87 @@
+            capabilities.eglServerWaitSupported = true;
+        } else {
+            capabilities.eglSupported = extensions.KHR_fence_sync;
+            capabilities.eglServerWaitSupported = extensions.KHR_fence_sync;
```
Shouldn't this be KHR_wait_sync?

carlosgcampos replied:
Indeed! good catch.

magomez commented on Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp:
```diff
@@ -0,0 +1,130 @@
+        Vector<EGLint> intAttributes = attributes.map<Vector<EGLint>>([] (EGLAttrib value) {
+            return value;
+        });
+        sync = eglCreateSyncKHR(display.eglDisplay(), type, intAttributes.isEmpty() ? nullptr : intAttributes.data());
```
This can cause a build error on platforms with EGL 1.4 that don't have the extension, as the KHR functions may not be defined.

carlosgcampos replied:
I don't think so, epoxy takes care of that.

magomez commented on Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp:
```diff
@@ -0,0 +1,130 @@
+    if (display.eglCheckVersion(1, 5))
+        eglDestroySync(display.eglDisplay(), m_sync);
+    else
+        eglDestroySyncKHR(display.eglDisplay(), m_sync);
```
Same as before.

magomez commented on Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp:
```diff
@@ -0,0 +1,130 @@
+    if (display.eglCheckVersion(1, 5))
+        eglClientWaitSync(display.eglDisplay(), m_sync, 0, EGL_FOREVER);
+    else
+        eglClientWaitSyncKHR(display.eglDisplay(), m_sync, 0, EGL_FOREVER_KHR);
```
Ditto

magomez commented on Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp:
```diff
@@ -0,0 +1,130 @@
+    if (display.eglCheckVersion(1, 5))
+        eglWaitSync(display.eglDisplay(), m_sync, 0);
+    else
+        eglWaitSyncKHR(display.eglDisplay(), m_sync, 0);
```
Ditto

carlosgcampos committed:
[GTK][WPE] Add support for EGL fences
https://bugs.webkit.org/show_bug.cgi?id=277141

Reviewed by Miguel Gomez.

For explicit sync support we need to use the EGL API to create the
fences to be able to use the EGL_ANDROID_native_fence_sync extension.
So, we can make it possible to create a GLFence with either gl or egl
APIs depending on what we need and what's available.
There's no change in behavior in this patch, but now that we have EGL
fences we prefer to use them when available if server wait is supported.

* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::initializeEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/egl/GLFence.cpp:
(WebCore::GLFence::capabilities):
(WebCore::GLFence::isSupported):
(WebCore::GLFence::create):
(WebCore::GLFence::createExportable):
(WebCore::GLFence::importFD):
(WebCore::GLFence::GLFence): Deleted.
(WebCore::GLFence::~GLFence): Deleted.
(WebCore::GLFence::wait): Deleted.
* Source/WebCore/platform/graphics/egl/GLFence.h:
(WebCore::GLFence::exportFD):
* Source/WebCore/platform/graphics/egl/GLFenceEGL.cpp: Added.
(WebCore::createEGLFence):
(WebCore::GLFenceEGL::create):
(WebCore::GLFenceEGL::createExportable):
(WebCore::GLFenceEGL::importFD):
(WebCore::GLFenceEGL::GLFenceEGL):
(WebCore::GLFenceEGL::~GLFenceEGL):
(WebCore::GLFenceEGL::clientWait):
(WebCore::GLFenceEGL::serverWait):
(WebCore::GLFenceEGL::exportFD):
* Source/WebCore/platform/graphics/egl/GLFenceEGL.h: Added.
* Source/WebCore/platform/graphics/egl/GLFenceGL.cpp: Added.
(WebCore::GLFenceGL::create):
(WebCore::GLFenceGL::GLFenceGL):
(WebCore::GLFenceGL::~GLFenceGL):
(WebCore::GLFenceGL::clientWait):
(WebCore::GLFenceGL::serverWait):
* Source/WebCore/platform/graphics/egl/GLFenceGL.h: Added.

Canonical link: https://commits.webkit.org/281558@main
